src/thread_operations.c

```diff
@@ -0,0 +1,83 @@
+#include <FreeRTOS.h>
+#include <semphr.h>
+#include <stdio.h>
+#include <pico/stdlib.h>
+#include <pico/multicore.h>
+#include <pico/cyw43_arch.h>
+#include <stdint.h>
+#include <FreeRTOS.h>
+#include <task.h>
+
+#include "thread_operations.h"
+
+bool side_thread_function(int *counter, SemaphoreHandle_t semaphore)
+{
+    bool semaphore_pass = xSemaphoreTake(semaphore, 500);
+    if(semaphore_pass)
+    {
+        *counter = *counter + 1;
+        xSemaphoreGive(semaphore);
+	    printf("hello world from %s! Count %d\n", "thread", *counter);
+    }
+   
+    return semaphore_pass;
+}
+
+bool main_thread_function(int *counter, int *on, SemaphoreHandle_t semaphore)
+{
+    bool semaphore_pass = xSemaphoreTake(semaphore, 500);
+    if (semaphore_pass)
+    {
+        printf("hello world from %s! Count %d\n", "main", (*counter)++);
+        xSemaphoreGive(semaphore);
+        *on = !(*on);
+    }
+    return semaphore_pass;
+}
+
+void thread_deadlock_function(void *params)
+{
+    struct DeadlockParams *dl_args = (struct DeadlockParams *)params;
+
+    while(!xSemaphoreTake(dl_args->sem_a, 500));
```

Comment:
There is a macro pdMAX_DELAY or something like that if you wanted to avoid the while loop.